### ISSUE TITLE
about bs of non-local resnet50 on Charades

### ISSUE THREAD

**Issue by ZilinGao** [gsig/PyVideoResearch#13]
Hi~ Thanks for your work.

The default bs of non-local resnet50 on Charades is 5. But following this setting, we(with @icyzhang0923 ) got an mAP with 29.98, which is an obvious gap with the report result 32.08. So we change the bs and get the following results:

|bs    |  gpus   |    bs/gpu  | mAP(epoch45)|
|:----:|:--------:|:-----------:|:---------------:|
|5     |          2  |           2.5  |   29.98     (default setting)|
|8      |    2         |    4             |     31.29|
|16    |    4         |    4             |    31.16 |                   
|32    |    8         |    4             |     29.18|
|??     |    ??       |     ??         |    32.08    (report)|

On account of your released model is on epoch47. So we evaluate epoch45 merely. Other settings are all consistent with default options and employed the pretrained model you provided. So, could you please tell us your bs setting and gpu number?

Thanks a lot.

**Comment by gsig** [comment]
Looking through the log file for nonlocal_resnet50_3d_charades.py: https://www.dropbox.com/s/8ri8gv0eytgzzrz/i3d31b.txt?dl=0

I noticed that there is indeed a difference between the experiment file nonlocal_resnet50_3d_charades.py and what it was trained with, it may have been because it was tested on a single gpu or something.. in any case the correct parameters are: (line 2 of the log file)
```
'name': 'i3d31b',
'resume': '/nfs.yoda/gsigurds/ai2/caches/i3d8l2/model_best.pth.tar',
'arch': 'resnet50_3d_nonlocal',
'input_size': 224,
'dropout': 0,
'pretrained': True,
'nclass': 157,
'wrapper': 'default_wrapper',
'criterion': 'default_criterion',
'replace_last_layer': True,
'workers': 12,
'disable_cudnn_benchmark': True,
'disable_cudnn': True,
'epochs': 50,
'start_epoch': 1,
'batch_size': 20,
'video_batch_size': 20,
'lr': 0.375,
'lr_decay_rate': '15,40',
'weight_decay': 1e-07,
'train_size': 1.0,
'val_size': 0.2,
'window_smooth': 0,
```

so the batch size indeed 20 (4 gpus) for 50 epochs with 0.375 learning rate and decay schedule '15,40'. The git hash of the repo that ran this experiment is 713cf90 in case you have trouble reproducing this with the most recent version of the repo.

Keep in mind that you can further increase the performance by increasing the image size during training or testing, and doing multi-crop testing etc etc, like the do in the nonlocal paper.

Let me know if that helps.

**Comment by ZilinGao** [comment]
Thanks for you response.

I rerun the experiment with bs20 and obtaining mAP with 31.42, still having gap with 32.08. 
Here is my script file:

  


    name = __file__.split('/')[-1].split('.')[0]

     args = [
    '--name', name,  # name is filename
    '--print-freq', '5',
    '--dataset', 'charades_video',
    '--arch', 'resnet50_3d_nonlocal',
    '--lr', '.375',
    '--criterion', 'default_criterion',
    '--wrapper', 'default_wrapper',
    '--lr-decay-rate', '15,40',
    '--epochs', '50',
    '--batch-size', '20',
    '--video-batch-size', '20',
    '--train-size', '1.0',
    '--weight-decay', '0.0000001',
    '--val-size', '0.2',
    '--cache-dir', 'caches/',
    '--data', '/home/gaozilin/data/Charades_v1_frame',
    '--train-file', 'notation_Charades/Charades_v1_train.csv',
    '--val-file', 'notation_Charades/Charades_v1_test.csv',
    '--pretrained',
    '--pretrained-weights','pretrained/i3d8l.pth.tar',
    '--start-epoch', '1',
    '--resume', 'caches/' + name + '/model.pth.tar' +
                '',
    '--workers', '48',
    '--disable-cudnn-benchmark',
    '--disable-cudnn',
    '--tasks', 'video_task',
    '--replace-last-layer',
    '--metric', 'video_task_CharadesmAP',
    ]

**Comment by gsig** [comment]
It's hard to say. It might just be random chance, or differences in the pytorch version etc etc. Have you tried pulling the exact version of the repo that was used to train that experiment? ( 713cf90 ) It's possible that something is slightly different that could cause this discrepancy and/or require slightly different parameters.

**Comment by ZilinGao** [comment]
The reruned experiment used latest code version. But it looks like the network and non-local layer are the same in the two  versions. So the result may be not change a lot.  I will rerun it with the version you mentioned when the machine is avaliable.

Thank you.

**Comment by ZilinGao** [comment]
Hello~ I rerun the non-local experiments on Charades with ResNet-50 backbone. The batch size is 18 on three gpus in total. The 50 clips mAP is 31.95 which is very close to report performance 32.08. So it seems that the code and hyper-parameter provided are no problem. Thanks for your nice code~
